### PR TITLE
Remove `_data` attribute from `DataSampler`

### LINKED ISSUE #327: Improve DataSampler efficiency
### Problem Description

To make CTGAN sampling more efficient, we can store the frequency of each value for every categorical variable and not store the `self._data` variable in the [DataSampler](https://github.com/sdv-dev/CTGAN/blob/39f896e817afd64a3f46dce74252fe196ffd74bd/ctgan/data_sampler.py).

### Expected behavior

We can change the code in the following places:

1. In the following lines, we can pass the data in from the [fit ](https://github.com/sdv-dev/CTGAN/blob/39f896e817afd64a3f46dce74252fe196ffd74bd/ctgan/synthesizers/ctgan.py#L281)call of CTGAN.
https://github.com/sdv-dev/CTGAN/blob/39f896e817afd64a3f46dce74252fe196ffd74bd/ctgan/data_sampler.py#L136-L138
2. In the lines below, the `_data` is used to pick a value for a random conditional variable and set it to 1 in the conditional vector. Instead of doing this, we can just store the frequency for each value of each conditional variable and sample accordingly from that.
https://github.com/sdv-dev/CTGAN/blob/39f896e817afd64a3f46dce74252fe196ffd74bd/ctgan/data_sampler.py#L120-L126
The code above is doing the following:
1. Picking a random row index
2. Picking a random categorical variable
3. Finding the start and end index corresponding to that categorical variable (This is because the data it stores is the transformed data and everything is 1 hot encoded with no column names so it has to figure out which columns correspond to the selected category)
4. Using the selected row index to extract a real row from the transformed data
5. Figure out which column is 1 (which value of the category that row was)
6. In an array of all 0s that is the same length as the row, set the value corresponding to the category found in step 5 to 1. That is the conditional vector that will be used

There is no reason to pick the whole row if we just need the value for the randomly selected discrete variable.

### PR DESCRIPTION
**Pull request by fealho**
CU-86az1bfrb, Resolve #327.

### CODE REVIEW

**Comment by sdv-team** [comment]
Task linked: [CU-86az1bfrb CTGAN - Improve DataSampler efficiency #327](https://app.clickup.com/t/86az1bfrb)

File: ctgan/data_sampler.py [hunk 1]
```diff
@@ -115,33 +115,30 @@ def sample_original_condvec(self, batch):
...
-            matrix_ed = matrix_st + self._discrete_column_n_category[col_idx]
-            pick = np.argmax(self._data[row_idx, matrix_st:matrix_ed])
-            cond[i, pick + self._discrete_column_cond_st[col_idx]] = 1
+        cond[np.arange(batch), col_idxs] = 1
 
         return cond
 
-    def sample_data(self, n, col, opt):
+    def sample_data(self, data, n, col, opt):
         """Sample data from original training data satisfying the sampled conditional vector.
```
Comment (rwedge) [inline ctgan/data_sampler.py:128]: should we update this description now that data is an input to the method?

File: ctgan/data_sampler.py [hunk 2]
```diff
@@ -115,33 +115,33 @@ def sample_original_condvec(self, batch):
...
 
         Returns:
-            n rows of matrix data.
+            data:
+                The training data.
+            n:
+                n rows of matrix data.
```
Comment (rwedge) [inline ctgan/data_sampler.py:134]: Should this be the Args documentation block?

**Review by frances-h — APPROVED** [review]

**Review by rwedge — APPROVED** [review]
left a minor documentation comment

**Pull request merged**